nameko/amqp.py

```diff
@@ -47,3 +51,33 @@ def verify_amqp_uri(amqp_uri):
     transport = TestTransport(connection.transport.client)
     with transport.establish_connection():
         pass
+
+
+class UndeliverableMessage(Exception):
+    """ Raised when publisher confirms are enabled and a message could not
+    be routed or persisted """
+    pass
+
+
+@contextmanager
+def get_connection(amqp_uri):
+    conn = Connection(amqp_uri)
+    with connections[conn].acquire(block=True) as connection:
+        yield connection
+
+
+@contextmanager
+def get_producer(amqp_uri, confirms=True):
+    transport_options = {
+        'confirm_publish': confirms
+    }
+    conn = Connection(amqp_uri, transport_options=transport_options)
+
+    with producers[conn].acquire(block=True) as producer:
+        yield producer
```

Comment:
Hmm. As it's implemented here you can send multiple undeliverable messages and be notified about only one of them, and then only after you exit the contextmanager.

It might be better to push this check down in to the different implementations.